### PR TITLE
Fix for issue #21

### PR DESCRIPTION
**Pull request by andr2000**
This fixes "arm/coproc: coproc wait_time logic has to be reviewed" #21

### CODE REVIEW

File: xen/arch/arm/coproc/coproc.h [hunk 1]
```diff
@@ -167,6 +165,30 @@ static inline void vcoproc_get_rw_context(struct domain *d, struct mmio *mmio,
...
+ *        coproc must be considered as non-functional
+ */
+static inline s_time_t vcoproc_context_switch_from(struct vcoproc_instance *vcoproc)
+{
+    return vcoproc->coproc->ops->ctx_switch_from(vcoproc);
+}
+
+/* new vcoproc has time slice, restore context and unpause:
+ * return value:
+ *  >=0 - on success
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/coproc.h]: Can't we just use 0 in case of success?

File: xen/arch/arm/coproc/schedule.c [hunk 2]
```diff
@@ -233,6 +233,53 @@ static inline void schedule_trace(struct vcoproc_instance *curr,
     }
 }
 
+static s_time_t vcoproc_schedule_context_switch(struct vcoproc_instance *curr,
+                                                struct vcoproc_instance *next)
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/schedule.c]: vcoproc_shedule_context_switch -> vcoproc_sheduler_context_switch

File: xen/arch/arm/coproc/schedule.c [hunk 3]
```diff
@@ -253,15 +253,20 @@ static s_time_t vcoproc_schedule_context_switch(struct vcoproc_instance *curr,
 
         wait_time = vcoproc_context_switch_from(curr);
 
-        if ( wait_time == 0 )
+        if ( likely(wait_time == 0) )
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/schedule.c]: No sure that we need likely here.

Comment (aanisov) [inline xen/arch/arm/coproc/schedule.c]: I guess wee expect that coprocessors would reject context_switch_from rarely.

File: xen/arch/arm/coproc/schedule.c [hunk 4]
```diff
@@ -258,7 +305,7 @@ void vcoproc_schedule(struct vcoproc_scheduler *sched)
         goto out;
     }
 
-    wait_time = vcoproc_context_switch(curr, next);
+    wait_time = vcoproc_schedule_context_switch(curr, next);
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/schedule.c]: vcoproc_schedule_context_switch -> vcoproc_scheduler_context_switch

File: xen/arch/arm/coproc/schedule.c (hunk 3 above)
Comment (aanisov) [inline xen/arch/arm/coproc/schedule.c]: As I understand (wait_time>0) is another `if` check, so it should be `likely()` as well keeping in mind that normally we do not expect `wait_time < 0`

File: xen/arch/arm/coproc/schedule.c [hunk 5]
```diff
@@ -253,15 +253,20 @@ static s_time_t vcoproc_schedule_context_switch(struct vcoproc_instance *curr,
...
                 curr->state = VCOPROC_SLEEPING;
         }
-        if ( wait_time )
+        else if ( wait_time > 0 )
             return wait_time;
+        else
+        {
+            panic("Failed to switch context from vcoproc \"%s\" (%d)\n",
+                  dev_path(coproc->dev), ret);
+        }
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/schedule.c]: Why do you print "ret"? It is uninitialized here.

File: xen/arch/arm/coproc/coproc.h (hunk 1 above)
Comment (andr2000) [inline xen/arch/arm/coproc/coproc.h]: no, b/c then you have to explain what if result is >0

File: xen/arch/arm/coproc/schedule.c (hunk 2 above)
Comment (andr2000) [inline xen/arch/arm/coproc/schedule.c]: done: vcoproc_scheduler_context_switch

File: xen/arch/arm/coproc/coproc.h (hunk 1 above)
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.h]: So you can explain in a following way:

> =0 is OK
> !=0 is not OK

File: xen/arch/arm/coproc/schedule.c (hunk 3 above)
Comment (andr2000) [inline xen/arch/arm/coproc/schedule.c]: will remove likely for now

File: xen/arch/arm/coproc/schedule.c (hunk 4 above)
Comment (andr2000) [inline xen/arch/arm/coproc/schedule.c]: done

File: xen/arch/arm/coproc/schedule.c (hunk 5 above)
Comment (andr2000) [inline xen/arch/arm/coproc/schedule.c]: done

File: xen/arch/arm/coproc/coproc.h (hunk 1 above)
Comment (otyshchenko1) [inline xen/arch/arm/coproc/coproc.h]: Sorry, but I don't understand. We have mismatch between code and comment. In comment you point that >=0 is success. In code you rely on =0 is success.

Comment (andr2000) [inline xen/arch/arm/coproc/coproc.h]: you usually check the return value for <0
why do you want this to be different in our case?

Comment (aanisov) [inline xen/arch/arm/coproc/coproc.h]: We usually check **status** to be equal 0. Time **value** is checked to be not valid on <0 in context_switch_from()

Comment (aanisov) [inline xen/arch/arm/coproc/coproc.h]: We usually check **status** value to be 0 for success. And we checked **time** value to be <0 in context_switch_from().

Comment (andr2000) [inline xen/arch/arm/coproc/coproc.h]: 1. The return value is status
2. I usually check status for <0 which in most cases is error
3. we are discussing static inline int vcoproc_context_switch_to(struct vcoproc_instance *vcoproc)

Comment (klogg) [inline xen/arch/arm/coproc/coproc.h]: We have a definition of function, so it is absolutely author's right to limit return value to `(-&infin; ; 0]` and so describe only those values.
```c
+ * return value:
+ *  0 - no error
+ *  <0 - error code
```
BTW, do we use standard error codes here? Just wondering.

**Comment by andr2000** [comment]
please review the updated version

**Comment by otyshchenko1** [comment]
Updated version looks brilliant

**Comment by andr2000** [comment]
updated

**Pull request merged**